lib/OpenQA/WebAPI/Plugin/AMQP.pm

```diff
@@ -0,0 +1,182 @@
+# Copyright (C) 2016 SUSE Linux GmbH
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, see <http://www.gnu.org/licenses/>.
+
+package OpenQA::WebAPI::Plugin::AMQP;
+
+use strict;
+use warnings;
+
+use parent qw/Mojolicious::Plugin/;
+use JSON;
+use Mojo::IOLoop;
+use OpenQA::Utils;
+use OpenQA::Schema::Result::Jobs;
+use Mojo::RabbitMQ::Client;
+
+my $app;
+my $config;
+my $client;
+my $channel;
+my $reconnecting = 0;
+
+my @job_events     = qw/job_create job_delete job_cancel job_duplicate job_restart job_update_result job_done/;
+my @comment_events = qw/comment_create comment_update comment_delete/;
+
+sub reconnect {
+    my $self = shift;
+
+    return if $reconnecting;
+    $reconnecting = 1;
+    Mojo::IOLoop->timer(
+        5 => sub {
+            $reconnecting = 0;
+            $self->connect();
+        });
+}
+
+sub connect {
+    my $self = shift;
+
+    OpenQA::Utils::log_debug("Connecting to AMQP server");
+    $config = $app->config;
+    $client = Mojo::RabbitMQ::Client->new(url => $config->{amqp}{url});
+    $client->catch(sub { OpenQA::Utils::log_warning("Failed to connect to AMQP server: " . $config->{amqp}{url}); });
+    $client->on(
+        open => sub {
+            my ($client) = @_;
+
+            $channel = Mojo::RabbitMQ::Channel->new();
+            $channel->catch(sub { OpenQA::Utils::log_warning("Error on AMQP channel received"); });
+
+            $channel->on(
+                open => sub {
+                    my ($channel) = @_;
+                    $channel->declare_exchange(exchange => $config->{amqp}{exchange}, type => 'topic')->deliver();
+                });
+            $channel->on(
+                close => sub {
+                    OpenQA::Utils::log_warning("AMQP channel closed");
+                });
+            $client->open_channel($channel);
+        });
+    $client->on(
+        close => sub {
+            OpenQA::Utils::log_warning("AMQP connection closed");
+            $self->reconnect();
+        });
+    $client->on(
+        error => sub {
+            OpenQA::Utils::log_warning("AMQP connection error");
+            $self->reconnect();
+        });
+    $client->on(
+        disconnect => sub {
+            OpenQA::Utils::log_warning("AMQP connection closed");
+            $self->reconnect();
```

Comment:
Won't this result in endless reconnecting if the server is not available? It might make sense to wait some time after a few failed attempts before continuing.